tianshou/data/utils/segtree.py

```diff
@@ -0,0 +1,122 @@
+    4. Query an interval [l, r] with the default operation takes O(log(bound))
+
+    :param int size: the size of segment tree.
+    :param str operation: the operation of segment tree. Choose one of "sum",
+        "min" and "max", defaults to "sum".
+    """
+
+    def __init__(self, size: int,
+                 operation: str = 'sum') -> None:
+        bound = 1
+        while bound < size:
+            bound *= 2
+        self._size = size
+        self._bound = bound
+        assert operation in ['sum', 'min', 'max'], \
+            f'Unknown operation {operation}.'
+        if operation == 'sum':
+            self._op, self._init_value = np.add, 0.
+        elif operation == 'min':
+            self._op, self._init_value = np.minimum, np.inf
+        else:
+            self._op, self._init_value = np.maximum, -np.inf
+        # assert isinstance(self._op, np.ufunc)
+        self._value = np.full([bound * 2], self._init_value)
+
+    def __len__(self):
+        return self._size
+
+    def __getitem__(self, index: Union[int, np.ndarray]
+                    ) -> Union[float, np.ndarray]:
+        """Return self[index]"""
+        return self._value[index + self._bound]
+
+    def __setitem__(self, index: Union[int, np.ndarray],
+                    value: Union[float, np.ndarray]) -> None:
+        """Insert or overwrite a (or some) value(s) in this segment tree. The
+        duplicate values are handled as numpy array, in other words, we only
+        keep the last value and ignore the previous same value.
+        """
+        if isinstance(index, int) and isinstance(value, float):
+            index, value = np.array([index]), np.array([value])
+        assert ((0 <= index) & (index < self._bound)).all()
+        index = index + self._bound
+        self._value[index] = value
+        while index[0] > 1:
+            index //= 2
+            self._value[index] = self._op(
+                self._value[index * 2], self._value[index * 2 + 1])
+
+    def reduce(self, start: Optional[int] = 0,
+               end: Optional[int] = None) -> float:
+        """Return operation(value[start:end])."""
+        if start == 0 and end is None:
+            return self._value[1]
+        if end is None:
+            end = self._bound
+        if end < 0:
+            end += self._bound
+        start, end = start + self._bound - 1, end + self._bound
+        result = self._init_value
+        while start ^ end ^ 1 != 0:
+            if start % 2 == 0:
+                result = self._op(result, self._value[start ^ 1])
+            if end % 2 == 1:
+                result = self._op(result, self._value[end ^ 1])
+            start, end = start // 2, end // 2
+        return result
+
+    def get_prefix_sum_idx(
+            self, value: Union[float, np.ndarray]) -> Union[int, np.ndarray]:
+        """Return the index ``i`` which satisfies
+        ``sum(value[:i]) <= value < sum(value[:i + 1])``.
```

Comment:
If there are multiple indexes that satisfy this, return the largest, smallest, or whichever one?

Reply:
This must be the smallest one. Suppose we are using a buffer with capacity of 100, and there are just three elements now, we have to assure the returned index is smaller than 3, otherwise we are trying to access an transition tuple that does not exist.